Implement more special values of polylogarithm, without exp_polar for s=1

normalhuman opened the pull request:
#### References to other Issues or PRs

Closes #7132, also closes #13853.

#### Brief description of what is fixed or changed

`polylog(s, z)` has simple formulas for s = 0 and s = -1, These are now directly used by eval, just as it uses the formulas for z in -1, 0, 1 without the user needing to expand_func. For example:
```
>>> polylog(0, 2)
-2    # previously this returned unevaluated
```
One has to draw the line somewhere, and I left s = -2 and the rest un-evaluated, subject to the user applying `expand_func`. Also, keeping `polylog(1, z)` in the zeta-function family is better for series summation methods in cases when such sums end up being represented as zeta-function.

Per the discussion at #13853 the function expansion for s = 1 is now simply `-log(1 - z)`, as it appears in Wikipedia, Wolfram, etc.

#### Dilogarithm

Expressions such as polylog(2, 2) and polylog(2, 1/2) now also evaluate in closed form. The values are confirmed by comparing [Wikipedia](https://en.wikipedia.org/wiki/Spence%27s_function#Special_values) and [Wolfram Functions](http://functions.wolfram.com/ZetaFunctionsandPolylogarithms/PolyLog2/03/02/) and additionally checked by floating point evaluation.

**Add entry(ies) to the release notes?** No

normalhuman committed:
Implement special values of dilogarithm, most notably at 2 and 1/2

Expressions such as polylog(2, 2) and polylog(2, 1/2) now evaluate
in closed form. Closes #7132.

The values are confirmed by comparing
 https://en.wikipedia.org/wiki/Spence%27s_function#Special_values
and
 http://functions.wolfram.com/ZetaFunctionsandPolylogarithms/PolyLog2/03/02/
and additionally checked by floating point evaluation.

jksuom commented on sympy/functions/special/tests/test_zeta_functions.py:
```diff
@@ -134,6 +134,13 @@ def test_polylog_expansion():
+    assert polylog(2, 2) == pi**2/4 - I*pi*log(2)
+    assert polylog(2, S.Half) == pi**2/12 - log(2)**2/2
+    for z in [S.Half, 2, (sqrt(5)-1)/2, -(sqrt(5)-1)/2, -(sqrt(5)+1)/2, (3-sqrt(5))/2]:
+        assert Abs(polylog(2, z) - polylog(2, z + 1e-13)).evalf() < 1e-12
```
The numeric value computed by `mpmath` can be obtained by setting `evaluate=False`.
```
>>> polylog(2, 2)
pi**2/4 - I*pi*log(2)
>>> polylog(2, 2, evaluate=False)
polylog(2, 2)
>>> polylog(2, 2, evaluate=False).evalf()
2.46740110027234 - 2.1775860903036*I  # this is computed by mpmath polylog
```

normalhuman committed:
Test polylog values by comparing with evaluations with evaluate=False

To make sure none of special polylog values are mistyped, they are compared to
what mpmath returns without going through SymPy eval method.

normalhuman committed:
Polylog should return S.Zero instead of Python's 0 when z=0

The returned value is expected to be a SymPy object, "return 0"
in a SymPy function is incorrect.

normalhuman committed:
Evaluate polylog(s, z) for s in [-1, 0, 1], without exp_polar

polylog(s, z) has simple formulas for s in -1, 0, 1. They should be
used by eval, just as it uses the formulas for z in -1, 0, 1 without
the user needing to expand_func. For example:
```
>>> polylog(2, S(1)/2)
-log(2)**2/2 + pi**2/12
>>> polylog(1, S(1)/2)
log(2)     # previously this returned unevaluated
```
One has to draw the line somewhere, and I left s = -2 and the rest
un-evaluated, subject to the user applying expand_func

Per the discussion at #13853 the expansion for s = -1 is simply
-log(1 - z), as it appears in Wikipedia, Wolfram, etc.

normalhuman committed:
Do not expand polylog(1, z) to -log(1-z) without expand_func

It seemed like a good idea, but series summation methods work better
with polylog when the sum is to be expressed in terms of zeta function.
